feat: Add example Carbon Black function

darrylcauldwell opened the pull request:
Closes: #866
Signed-off-by: Darryl Cauldwell <darryl.cauldwell@gmail.com>

## Summary

Please describe your changes and what you plan to introduce as part of this pull request submission. If it fixes a bug or resolves a feature request, be sure to link to the issue numbers following the template below.

We are looking to trigger a function when VmCreatedEvent event occurs which trigger software installation once guest OS is booted AND VMware Tools is running.  The time between event and state where software install can be performed will depend on VM startup time.

## Pull Request Checklist
🚨 Please review the [guidelines for contributing](https://vmweventbroker.io/community) to this repository.

- [X] Please ensure that you are making a pull request against the **Development** branch
- [X] Please use the `WIP` keyword in the title of your PR if you are not ready for review
- [X] Please ensure that you have opened a Github Issue if you are resolving/fixing a problem
- [X] Please ensure that you have [signed](https://help.github.com/en/github/authenticating-to-github/signing-commits) all commits and that you have [squashed](https://medium.com/@slamflipstrom/a-beginners-guide-to-squashing-commits-with-git-rebase-8185cf6e62ec) all relevant commits related to your change
- [X] Please make sure that you have tested your change locally by successfully [building and deploying the VMware Event Broker Appliance](https://vmweventbroker.io/kb/contribute-appliance) and/or [building and deploying VMware Event Router](https://vmweventbroker.io/kb/contribute-eventrouter)
- [X] Please include any relevant screenshots and/or output as part of your testing
- [X] Please include any documentation updates that is applicable for your changes

## Change Type

What types of changes does your code introduce to the VMware Event Broker Appliance?

_Put an `x` in all boxes that apply_

Please check the type of change your PR introduces:
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] Feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation changes
- [ ] Other (please describe):

## Resolved Issues

List of Issues closed or resolved by this PR. Add multiple `Closes` keyword followed by the issue number (e.g. Closes #ISSUE-NUMBER)

Closes #ISSUE-NUMBER

## Testing Verification

* Short summary of testing (e.g. successfully built and deployed VMware Event Broker Appliance)
* Please include any relevant screenshots and/or output as part of the testing

## Additional Information

* Any other details you wish to include or mention

If you have any questions/comments, feel free to reach out to team on Slack [#vcenter-event-broker-appliance](https://vmwarecode.slack.com/archives/CQLT9B5AA)

Thank you from the VEBA Team! 🥳

embano1 commented:
@darrylcauldwell not sure what happened but you opened a new PR and looks like some (most?) of the feedback from the previous PR (#885) hasn't been incorporated. 

Could you please verify? Also, not sure if you intended to have certain CB URLs/company IDs in the examples/ini files. Those should be faked, e.g. `replace-me`.

darrylcauldwell commented:
Yes you are correct,  I have just updated to remediate previous PR comments
- Added syntax highlighting to filenames within README.md.
- Changed <container-registry> to <docker-username>.
- It was intended that the app logger level reads .info and not .debug, logging outputs meaningful messages at INFO and increased verbose messages at DEBUG so should only be turned on when debugging not as default.
- The CompanyCode wasn’t sensitive as test org but changed as this would be customer specific.
- Added exception around SlackWebhook.
- Fixed typos in instance and immediately.
- Updated all functions to NOT scale to zero and run min=1.

darrylcauldwell commented:
Will look to squash changes into commit and let you know when complete.

embano1 commented on examples/knative/python/kn-py-cbc-slack/cbc-envs:
```diff
@@ -0,0 +1,5 @@
+CBC_URL=<ENTER YOUR CBC URL eg https://defense-prod05.conferdeploy.net>
```
Please remove any real IPs/DNS names

embano1 commented on examples/knative/python/kn-py-cbc-slack/README.md:
```diff
@@ -0,0 +1,105 @@
+
+The use cases for these are slightly different so we have included functions to deply them seperately or together.  The `function-deploy.yaml` would be used in situations where new virtual machines are being created using powercli for example and Carbon Black is not pre installed, in this scenario we are automating the deployment of a Carbon Black sensor whenever a new VM is deployed.  The `funct ...
+
+If you are making changes to the `handler.py` logic or want to host your own images in a private registry then you will need to recreate the images as below in steps 1 to 4.  If you do not need to make changes and are happy deploying the latest public image skip straight to Step 4.
```
```suggestion
If you are making changes to the `handler.py` logic or want to host your own images in a private registry then you will need to recreate the images as below in steps 1 to 4.  

**If you do not need to make changes and are happy deploying the latest public image skip straight to Step 3.**
```

embano1 commented on examples/knative/python/kn-py-cbc-slack/cbc-ini:
```diff
@@ -0,0 +1,6 @@
+[customer]
+EncodedCompanyCode = <ENTER YOUR CBC EncodedCompanyCode>
+CompanyCode = <ENTER YOUR CBC CompanyCode>
+BackendServer = dev-prod05.conferdeploy.net
```
see earlier comment

embano1 commented on examples/knative/python/kn-py-cbc-slack/cbc-ini:
```diff
@@ -0,0 +1,6 @@
+EncodedCompanyCode = <ENTER YOUR CBC EncodedCompanyCode>
+CompanyCode = <ENTER YOUR CBC CompanyCode>
+BackendServer = dev-prod05.conferdeploy.net
+ProxyServer = 192.168.200.75:3128
```
question: is this always needed? can you elaborate on this setting in the README perhaps?

embano1 commented on examples/knative/python/kn-py-cbc-slack/cbc-ini:
```diff
@@ -0,0 +1,6 @@
+CompanyCode = <ENTER YOUR CBC CompanyCode>
+BackendServer = dev-prod05.conferdeploy.net
+ProxyServer = 192.168.200.75:3128
+CurlCrlCheck = false
```
question: what does this do? any doc refs/link for these settings?

embano1 commented on examples/knative/python/kn-py-cbc-slack/cbc-url:
```diff
@@ -0,0 +1 @@
+<ENTER YOUR CBC URL eg https://defense-prod05.conferdeploy.net>
```
see earlier comment

embano1 commented on examples/knative/python/kn-py-cbc-slack/deploy/handler.py:
```diff
@@ -0,0 +1,98 @@
+    try:
+        slackWebhook = WebhookClient(url)
+    except:
+        return "Could not for Webhook to Slack"
```
is the message worded correctly?

embano1 commented on examples/knative/python/kn-py-cbc-slack/deploy/handler.py:
```diff
@@ -0,0 +1,98 @@
+                    return "", 202
+                else:
+                    app.logger.debug(f"VM not eligible for sensor installation,  retrying " + vm.eligibility)
+                    time.sleep(30)
```
nit: we should reduce this to say 10s because this is a blocking call and the cloudevent sender (broker) might not wait that long and consider your function to be dead.
